---
title: "CEMiTool_Hib"
author: "Kdu Trufen"
date: "21 de agosto de 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

# Set working directory
```{r}
setwd("~/Dropbox/Hib")
```

# get_Attribute_Field
```{r}
get_Attribute_Field = function (x, field, attrsep = "; ") 
{
    s = strsplit(x, split = attrsep, fixed = TRUE)
    sapply(s, function(atts) {
        a = strsplit(atts, split = "=", fixed = TRUE)
        m = match(field, sapply(a, "[", 1))
        if (!is.na(m)) {
            rv = a[[m]][2]
        }
        else {
            rv = as.character(NA)
        }
        return(rv)
    })
}
```

# Protein list
```{r, message=FALSE}
library(tidyverse)
Hib_proteins <- read_delim("~/Dropbox/Hib_Data/Haemophilus_influenzae_10810.GCF_000210875.1.proteintable.txt", delim = "\t", escape_double = FALSE, trim_ws = TRUE) %>% as.data.frame
```

# Gene list
```{r, message=FALSE}
library(ballgown)
library(tidyverse)
RefSeq_gff.genes <- gffRead("~/Dropbox/Hib_Data/GCF_000210875.1_ASM21087v1_genomic.gff") %>% filter(feature == "gene") %>% mutate(locustag = get_Attribute_Field(attributes, "Name", attrsep = ";")) %>% mutate(ID = get_Attribute_Field(attributes, "ID", attrsep = ";")) %>% mutate(OldLocusTag = get_Attribute_Field(attributes, "old_locus_tag", attrsep = ";")) %>% mutate(Gene_length = end-start) %>% full_join(Hib_proteins) %>% select(Name = locustag, OldLocusTag, product)
```

# Read in Hib Matrix form Rsubread alignment
```{r }
#Hib_CountData = read.csv("~/Dropbox/Hib_Data/Hib_Hisat2_featureCounts_table_countMultiMappingReads.txt", header=TRUE, sep = "\t", row.names = 1)
library(tidyverse)
# Count matrix v1 (Hisat k -5, featureCounts -M)
# Hib_CountData <- read_delim("~/Dropbox/Hib_Data/Hib_Hisat2_featureCounts_table_countMultiMappingReads.txt", "\t", escape_double = FALSE, trim_ws = TRUE, skip = 1) %>% as.data.frame

# Count matrix v2 (Hisat k -2, featureCounts )
# Hib_CountData <- read_delim("~/Dropbox/Hib_Data/2_featureCounts_table_2018_09_14.txt", "\t", escape_double = FALSE, trim_ws = TRUE, skip = 1) %>% as.data.frame

# Count matrix v3 (Hisat k -5, featureCounts -M --primary)
Hib_CountData <- read_delim("~/Dropbox/Hib_Data/Hib_Hisat2_featureCounts_primary_table.txt", "\t", escape_double = FALSE, trim_ws = TRUE, skip = 1) %>% as.data.frame

library(stringr)
# Remove rrf gene (ribosomal RNA)
# set Geneid as row names
# remove unwanted columns
# change column names
# countData = Hib_CountData %>% filter(!str_detect(Geneid, pattern = "rrf")) %>% column_to_rownames('Geneid') %>% select(-c(Chr, Start, End, Strand, Length)) %>% setNames(c("S01B01", "S01B02", "S01B03", "S01B04", "S02B01", "S02B02", "S02B03", "S02B04", "S03B01", "S03B02", "S03B03", "S03B04", "S04B01", "S04B02", "S04B03", "S04B04", "S05B01", "S05B02", "S05B03", "S05B04", "S06B01", "S06B02", "S06B03", "S06B04"))

Hib_CountData = Hib_CountData %>% filter(!str_detect(Geneid, pattern = "rrf")) %>% column_to_rownames('Geneid') %>% select(-c(Chr, Start, End, Strand, Length)) %>% setNames( sort( as.vector( outer( paste0("S0", 1:6), paste0("B0", 1:4), paste0 ) ) ) )

Hib_CountData$S03B02 = NULL

THRESHOLD <- 0.01
LFC = 1
```

# Rename row names to old locus tag
```{r }
# Rename Hib genes to old locus tag (available at KEGG)
for (i in 1:nrow(RefSeq_gff.genes)){
  if (is.na(RefSeq_gff.genes$OldLocusTag[i])=="FALSE"){
    rownames(Hib_CountData)[which(rownames(Hib_CountData)==RefSeq_gff.genes$Name[i])] = RefSeq_gff.genes$OldLocusTag[i]
  }
}

# Sort row names alphabetically
# removeribosomal RNA
Hib_CountData = Hib_CountData %>% rownames_to_column %>% arrange(rowname) %>% filter(!str_detect(rowname, pattern = "HIB_r")) %>% column_to_rownames('rowname')
```

# Sets design
```{r }
# countData = Hib_CountData
HibDesign = data.frame(row.names = colnames( Hib_CountData ), condition = str_sub( colnames(Hib_CountData), start = 1L, end = 3L ) )

group=as.factor(HibDesign$condition)
design <- model.matrix(~0 + group)
colnames(design) <- levels(group)
rownames(design) = colnames(Hib_CountData)
```

```{r}
plot_genes_boxplot = function(df, genes, design, levels, converter, convert_from, convert_to){
  df = df %>% cpm %>% as.data.frame %>% rownames_to_column %>% filter(rowname %in% c(genes)) %>% distinct(rowname, .keep_all = TRUE) %>% arrange(rowname) %>% column_to_rownames
  if (nrow(df)%%4==0){
    df_list = lapply(1:floor(length(genes)/4), function(i) df[(4*i-3):(4*i),] )
  } else {
    df_list = lapply(1:floor(length(genes)/4), function(i) df[(4*i-3):(4*i),] )
    df_list[[(length(df_list)+1)]] = df[((length(df_list)*4)+1):length(genes),]
  }
  df_list = df_list %>% lapply(t) %>% lapply(function(x) as.data.frame(cbind(x, "Condition" = as.character(design, levels = levels)) ) )
  df_list = df_list %>% lapply(function(x) { x %>% gather(Gene, CPM, -Condition) } ) %>% lapply(function(x) { merge(x, converter, by.x = "Gene", by.y = convert_from) } )
  appender <- function(string, suffix = converter[[convert_to]][match(string, converter[[convert_from]])]) paste0(string, "\n (", suffix,")")
  
    lapply(df_list, function(x) ggplot(x, aes(x = Condition, y = as.numeric(CPM), color = Condition)) + geom_boxplot(alpha = .5) + geom_jitter(alpha = .5, width = .1) + theme_bw() + theme(axis.text.x = element_text(angle=90, hjust=1)) + ylab(label = "Read counts (cpm)") + theme(strip.text = element_text(size=10)) + facet_wrap(~Gene, scales = "free", ncol = 2, labeller = as_labeller(appender) ) )
    
  # lapply(df_list, function(x) ggplot(x, aes(x = Condition, y = as.numeric(CPM), color = Condition)) + geom_boxplot(alpha = .5) + geom_jitter(alpha = .5,width = .1) + theme_bw() + theme(axis.text.x = element_text(angle=90, hjust=1)) + ylab(label = "Read counts (cpm)") + facet_wrap(~Gene, scales = "free", ncol = 2, labeller = as_labeller( wrapit( appender) ) ) )
  
  # lapply(df_list, function(x) ggplot(x, aes(x = Condition, y = as.numeric(CPM), color = Condition)) + geom_boxplot(alpha = .5) + geom_jitter(alpha = .5,width = .1) + theme_bw() + theme(axis.text.x = element_text(angle=90, hjust=1)) + ylab(label = "Read counts (cpm)") + facet_wrap(~Gene, scales = "free", ncol = 2, labeller = label_wrap_gen( as_labeller(appender), width = 10) ) )
}

# swr = function(string, nwrap=20) {   paste(strwrap(string, width=nwrap), collapse="\n") } %>% Vectorize
wrapit <- function(text) {
  wtext <- paste(strwrap(text,width=10),collapse=" \n ")
  return(wtext)
}

genes = c("HIB_11840", "HIB_11830", "HIB_11820", "HIB_11810", "HIB_11690", "HIB_11680", "HIB_11670", "HIB_11660", "HIB_11730", "HIB_11720", "HIB_11710", "HIB_11870", "HIB_11860", "HIB_11850", "HIB_11650", "HIB_11800", "HIB_11640", "HIB_11790")

genes = c("HIB_11840", "HIB_11830", "HIB_11690", "HIB_11680")

plot_genes_boxplot(df = Hib_CountData, genes = genes, design = HibDesign$condition, levels = c("S01", "S02", "S03", "S04", "S05", "S06"), converter = RefSeq_gff.genes, convert_from = "OldLocusTag", convert_to = "product")

# appender(string = "HIB_11840", suffix = RefSeq_gff.genes[["product"]][match("HIB_11840", RefSeq_gff.genes[["OldLocusTag"]])])
# 
# label_wrap_gen(width=10)
```

# Filter
```{r }
library(edgeR)

countData = Hib_CountData

d <- DGEList(counts=countData, lib.size = colSums(countData), group=group)
# myCPM = cpm(d)
# cpm(1, mean(d$samples$lib.size))
# CPMthershold = round(cpm(1, mean(d$samples$lib.size)), digits=2)
# keep <- rowSums(cpm(d)> as.vector(CPMthershold)) >= 4
keep <- rowSums(cpm(d)> 1) >= 1
filtered <- d[keep,]

countData = filtered$counts
```

# least significantly DE genes based on a first-pass DE analysis performed prior to RUVg normalization.
```{r, message=FALSE}
library("edgeR")
library("RUVSeq")

lrt <- DGEList(counts=countData, lib.size = colSums(countData), group=group) %>% calcNormFactors(method="TMM") %>% estimateDisp(design, tagwise=TRUE, robust=TRUE) %>% glmFit(design) %>% glmLRT(coef=2:6)
top <-   topTags(lrt, n=nrow(d))$table
empirical <-   rownames(countData)[which(!(rownames(countData) %in% rownames(top)[1:500]))]
```

# Here, we consider all but the top 500 genes as ranked by edgeR p-values
```{r }
# The RUVg function  returns  two  pieces  of  information:
# the estimated factors of unwanted variation nd the normalized counts obtained by regressing the original counts on the unwanted factors
# The normalized values are stored in the normalizedCounts slot
set2 <- RUVg(as.matrix(d), empirical, k=1)
```

# plotRLE creates relative log expression (RLE) plot, initially proposed to measure the overall quality of a dataset 
# plotRLE can also be used to visualize the presence of unwanted batch effects in the data
```{r, fig.height = 5, fig.width = 10}
# mar - A numeric vector of length 4, which sets the margin sizes in the following order: bottom, left, top, and right. The default is c(5.1, 4.1, 4.1, 2.1).
# mgp - A numeric vector of length 3, which sets the axis label locations relative to the edge of the inner plot window. The first value represents the location the labels (i.e. xlab and ylab in plot), the second the tick-mark labels, and third the tick 
# las - A numeric value indicating the orientation of the tick mark labels and any other text added to a plot after its initialization. The options are as follows: always parallel to the axis (the default, 0), always horizontal (1), always perpendicular to the axis (2), and always vertical (3).

col.cell <- c("purple","orange", "blue", "red", "green", "black")[as.factor(HibDesign[,1])]
# plotRLE creates relative log expression (RLE) plot, initially proposed to measure the overall quality of a dataset 
# plotRLE can also be used to visualize the presence of unwanted batch effects in the data
par(mfrow=c(1,2), mar=c(8.1,4.1,4.1,2.1))
plotRLE(as.matrix(filtered$counts), outline=FALSE, ylim=c(-2.5, 2.5), col=col.cell, main = "Samples with \n unwanted variation", cex.main = 2, las=2, cex.axis = 0.8, style = "full",
       outlier.alpha = 0.1, outlier.shape = 3, outlier.size = 0, legend = TRUE)
legend("topright", inset = c(0,0), legend = levels(group), col=unique(col.cell), ncol=2, cex=0.5, border = "black", fill = unique(col.cell))
mtext(side = 1, text = "Samples", line = 7, cex = 2)
mtext(side = 2, text = "Relative Log Expression (RLE)", line = 1.5, cex = 2)

plotRLE(set2$normalizedCounts, outline=FALSE, ylim=c(-2.5, 2.5), col=col.cell, main = "Samples without \n unwanted variation", cex.main = 2, las=2, cex.axis = 0.8, style = "full",
       outlier.alpha = 0.1, outlier.shape = 3, outlier.size = 0)
legend("topright", inset = c(0,0), legend = levels(group), col=unique(col.cell), ncol=2, cex=0.5, border = "black", fill = unique(col.cell))
mtext(side = 1, text = "Samples", line = 7, cex = 2)
# mtext(side = 2, text = "Relative Log Expression (RLE)", line = 1.5, cex = 2) 
```

# Set design, removing batch effect
```{r}
# HibDesign2 = data.frame(row.names = colnames( countData ), condition = rep(c("S01", "S02", "S03", "S04", "S05", "S06"), each = 4),  set2$W)
# design = model.matrix(~group+set2$W, data = HibDesign2)
# colnames(design) <- c(levels(group), "W")
# group=as.factor(HibDesign2[,1])
```

#Pre-processing
```{r}
# We do not recommend filtering genes by differential expression. WGCNA is designed to be an unsupervised analysis method that clusters genes based on their expression profiles. Filtering genes by differential expression will lead to a set of correlated genes that will essentially form a single (or a few highly correlated) modules. It also completely invalidates the scale-free topology assumption, so choosing soft thresholding power by scale-free topology fit will fail. 
```

# TMM normaliztion from edgeR
```{r}
library("edgeR")
d <- DGEList(counts=set2$normalizedCounts, lib.size = colSums(set2$normalizedCounts), group=group)
WGCNAtpm <- calcNormFactors(d , "TMM") %>% cpm(normalized.lib.sizes=TRUE) %>% as.data.frame
head(WGCNAtpm)
```

```{r }
countData = set2$normalizedCounts
```

# sample_annot
```{r}
sample_annotation = data.frame("SampleName" = rownames(HibDesign), "Class" = HibDesign$condition)
```

```{r }
library(CEMiTool)

results <- new('CEMiTool', expression=WGCNAtpm, sample_name_column=sample_annotation$SampleName, class_column=sample_annotation$Class)

results <- find_modules(results,
                        cor_method='pearson',
                        cor_function='cor',
                        eps=0.1,
                        min_ngen=20,
                        merge_similar=TRUE,
                        diss_thresh=0.75,
                        network_type='unsigned',
                        tom_type='signed',
                        set_beta=NULL,
                        force_beta=TRUE,
                        verbose=FALSE)
```

```{r}
cem = results
unclassed = unclass(cem)
genes = attr(unclassed, "selected_genes")
# gene_symbols = unique(Conversion_Human2Human$SYMBOL[match(genes,Conversion_Human2Human$ENSEMBL)]) 
```

```{r}
nmodules(cem)
```

```{r}
head(module_genes(cem))
```

# Expression patterns in modules
```{r}
setGeneric('plot_profile', function(cem, ...) {
    standardGeneric('plot_profile')
})

#' @rdname plot_profile
setMethod('plot_profile', signature('CEMiTool'),
    function(cem, order_by_class=TRUE, center_func='mean') {
        if(!tolower(center_func) %in% c("mean", "median")){
            stop("Invalid center_func type. Valid values are 'mean' and 'median'")
        }
        modules <- unique(cem@module$modules)
        if(is.null(modules)){
               stop("No modules in this CEMiTool object.")
        }
        #vars <- mget(ls())
        #vars$modules <- NULL
        #cem <- get_args(cem=cem, vars=vars)

        modules <- modules[order(as.numeric(stringr::str_extract(modules, "\\d+")))]
        expr <- expr_data(cem)
        annot <- sample_annotation(cem)
        sample_name_column <- cem@sample_name_column
        class_column <- cem@class_column
        mod_cols <- mod_colors(cem)
        plots <- lapply(modules, function(mod){
            # subsets from expr all genes inside module mod
            genes <- cem@module[cem@module[,'modules']==mod, 'genes']
            expr[, 'id'] <- rownames(expr)
            mod_expr <- data.table::melt(expr[genes,], 'id',
                              variable.name='sample',
                              value.name='expression')

            # initialize plot base layer
            g <- ggplot(mod_expr, aes_(x=~sample, y=~expression))

            # adds different background colours if annot is provided
            if (nrow(annot)!=0) {
                if (order_by_class) {
                    # sorts data.frame by class name
                    annot <- annot[order(annot[, class_column]),]
                }
                annot[, sample_name_column] <- factor(annot[, sample_name_column],
                                                      levels=annot[, sample_name_column])
                mod_expr[, 'sample'] <- factor(mod_expr[, 'sample'],
                                               levels=annot[, sample_name_column])

                # y positioning of background tiles
                y_pos <- mean(mod_expr[, 'expression'])

                # reinitialize base layer adding background tiles
                g <- ggplot(mod_expr, aes_(x=~sample, y=~expression)) +
                     geom_tile(data=annot, alpha=0.3, height=Inf,
                               aes(x=get(sample_name_column), y=y_pos,
                               fill=as.factor(get(class_column))))
            }

            # adding lines
            g <- g + geom_line(aes_(group=~id), alpha=0.2, colour=mod_cols[mod]) +
                stat_summary(aes(group=1), size=1, fun.y=get(tolower(center_func)), geom='line')

            # custom theme
            g <- g + theme(plot.title=element_text(lineheight=0.8,
                                                   face='bold',
                                                   colour='black',
                                                   size=15),
                           axis.title=element_text(face='bold',
                                                   colour='black',
                                                   size=15),
                           axis.text.y=element_text(angle=0,
                                                    vjust=0.5,
                                                    size=12),
                           axis.text.x=element_text(angle=90,
                                                    vjust=0.5,
                                                    size=12),
                           panel.grid=element_blank(),
                           legend.title=element_blank(),
                           legend.text=element_text(size = 8),
                           legend.background=element_rect(fill='gray90',
                                                          size=0.5,
                                                          linetype='dotted'),
                           legend.position='bottom'
                           )
            # title
            g <- g + ggtitle(mod)

            return(g)
        })
        names(plots) <- modules
        cem@profile_plot <- plots
        return(cem)
})
```


# Expression patterns in modules
```{r}
plotprofile <- plot_profile(cem)
plots <- show_plot(plotprofile, "profile")

show_plot(plotprofile, "profile")
```

# plot gsea function from CEMiTool
```{r}
setGeneric('plot_gsea', function(cem, ...) {
    standardGeneric('plot_gsea')
})

#' @rdname plot_gsea
setMethod('plot_gsea', signature('CEMiTool'),
    function(cem, pv_cut=0.05) {
        if(length(unique(cem@module$modules)) == 0){
            stop("No modules in CEMiTool object! Did you run find_modules()?")
        }
        if(length(cem@enrichment) == 0){
            stop("No GSEA data! Did you run mod_gsea()?")
        }
        #cem <- get_args(cem, vars=mget(ls()))

        stats <- names(cem@enrichment)
        enrichment <- lapply(cem@enrichment, function(stat){
                                 stat[is.na(stat)] <- 0
                                 rownames(stat) <- stat[,1]
                                 stat[,1] <- NULL
                                 return(stat)
        })
        names(enrichment) <- stats

        pval <- enrichment[['pval']]
        nes <- enrichment[['nes']]

        pval <- pval[rowSums(pval < pv_cut) >= 1, , drop=FALSE]
        nes <- nes[rownames(pval), , drop=FALSE]

        # check if there is any signif. module
        if(nrow(nes) < 0){
            stop("No significant modules found!")
        }

        custom_pal <- c("#053061", "#2166AC", "#4393C3", "#92C5DE",
                        "#D1E5F0", "#FFFFFF", "#FDDBC7", "#F4A582",
                        "#D6604D", "#B2182B", "#67001F")
        custom_pal <- colorRampPalette(custom_pal)(200)

        nes <- as.matrix(nes)
        pval <- as.matrix(pval)
        nes[which(pval > pv_cut, arr.ind=TRUE)] <- 0

        if(nrow(nes) > 2){
            row_order <- rownames(nes)[hclust(dist(nes))$order]
        } else {
            row_order <- rownames(nes)
        }

        nes_melted <- reshape2::melt(nes)
        colnames(nes_melted) <- c("Module", "Class", "NES")
        nes_melted$Module <- factor(nes_melted$Module, levels=row_order)
        max_abs_nes <- max(abs(nes_melted$NES))
        res <- ggplot(nes_melted, aes_(x=~Class, y=~Module, size=~abs(NES), fill=~NES)) +
            geom_point(color = "white", shape=21) +
            scale_fill_gradientn(colours=custom_pal, space = "Lab",
                                 limits=c(-max_abs_nes, max_abs_nes)) +
            scale_size(range=c(0,9)) +
            guides(size="none") +
            theme_minimal() +
            theme(panel.grid.major = element_blank()) +
            scale_x_discrete(position = "top")
        res_list <- list(enrichment_plot=res)
        cem@enrichment_plot <- res_list

        return(cem)
})
```

# Module enrichment
```{r}
# When sample annotation is provided, the cemitool function will automatically evaluate how the modules are up or down regulated between classes. This is performed using the gene set enrichment analysis function from the fgsea package.

# You can generate a plot of how the enrichment of the modules varies across classes with the plot_gsea function. The size and intensity of the circles in the figure correspond to the Normalised Enrichment Score (NES), which is the enrichment score for a module in each class normalised by the number of genes in the module. This analysis is automatically run by the cemitool function, but it can be independently run with the function mod_gsea(cem)

sample_annotation(cem, 
                  sample_name_column="SampleName", 
                  class_column="Class") <- sample_annotation

gsea <- mod_gsea(cem)
gsea <- plot_gsea(gsea)

# gsea <- mod_gsea(cem) %>% plot_gsea(gsea)

show_plot(gsea, "gsea")
```

# KEGG
```{r}
library(clusterProfiler)
library(tidyverse)

# KEGG Enrichment Analysis of a gene set. Given a vector of genes, this function will return the enrichment KEGG categories with FDR control.

modules_genes_Hib = module_genes(cem) %>% arrange(modules) %>% split(f = .$modules) %>% lapply(select, genes) %>% lapply(pull)

Hib_KEGG_enrichment = modules_genes_Hib %>% compareCluster(fun="enrichKEGG", organism = 'hiu', pvalueCutoff=0.05)
modulesenrichKEGG = Hib_KEGG_enrichment %>% as.data.frame

#write.table( modulesenrichKEGG, paste0(path_to_files,"Hib_modulesenrichKEGGE_Hib.txt" ), sep="\t", row.names = FALSE, col.names = TRUE )

plot_Hib_KEGG_enrichment = Hib_KEGG_enrichment %>% plot(type="dot", font.size = 10, includeAll=TRUE, showCategory=10)
```

```{r }

```

```{r }

```

```{r }

```

```{r }

```

```{r }

```

```{r }

```

```{r }

```

```{r }

```

```{r, echo=FALSE}

```

