---
title: "Hib_edgeR_cpm_test"
author: "Kdu Trufen"
date: "2 de setembro de 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

# Set dir
```{r, message = FALSE}
rm(list=ls())
setwd("~/Dropbox/Hib/")
```

```{r, message = FALSE}
detachAllPackages <- function() {
    
    basic.packages <- c("package:stats","package:graphics","package:grDevices","package:utils","package:datasets","package:methods","package:base")
    
    package.list <- search()[ifelse(unlist(gregexpr("package:",search()))==1,TRUE,FALSE)]
    
    package.list <- setdiff(package.list,basic.packages)
    
    if (length(package.list)>0)  for (package in package.list) detach(package, character.only=TRUE)
    
}

detachAllPackages()
```

# Read in Hib Matrix form Rsubread alignment
```{r }
#Hib_CountData = read.csv("~/Dropbox/Hib_Data/Hib_Hisat2_featureCounts_table_countMultiMappingReads.txt", header=TRUE, sep = "\t", row.names = 1)
library(readr)
Hib_CountData <- as.data.frame(read_delim("~/Dropbox/Hib_Data/Hib_Hisat2_featureCounts_table_countMultiMappingReads.txt", "\t", escape_double = FALSE, trim_ws = TRUE, skip = 1))



library(stringr)
# Remove rrf gene (ribosomal RNA)
Hib_CountData = Hib_CountData[-which(str_detect(Hib_CountData$Geneid, "rrf")==TRUE),]

rownames(Hib_CountData) = Hib_CountData$Geneid
Hib_CountData = Hib_CountData[,-c(1:6)]
colnames(Hib_CountData) = c("S01B01", "S01B02", "S01B03", "S01B04", "S02B01", "S02B02", "S02B03", "S02B04", "S03B01", "S03B02", "S03B03", "S03B04", "S04B01", "S04B02", "S04B03", "S04B04", "S05B01", "S05B02", "S05B03", "S05B04", "S06B01", "S06B02", "S06B03", "S06B04")
countData = Hib_CountData

THRESHOLD <- 0.05
LFC = 1
```

# Sets design
```{r }
library('ballgown')
RefSeq_gff <- gffRead("~/Dropbox/Hib_Data/GCF_000210875.1_ASM21087v1_genomic.gff")
RefSeq_gff$Name <- getAttributeField(RefSeq_gff$attributes, "Name", attrsep = "; ")
RefSeq_gff$ID <- getAttributeField(RefSeq_gff$attributes, "ID", attrsep = "; ")
colnames(RefSeq_gff)

get_Attribute_Field = function (x, field, attrsep = "; ") 
{
    s = strsplit(x, split = attrsep, fixed = TRUE)
    sapply(s, function(atts) {
        a = strsplit(atts, split = "=", fixed = TRUE)
        m = match(field, sapply(a, "[", 1))
        if (!is.na(m)) {
            rv = a[[m]][2]
        }
        else {
            rv = as.character(NA)
        }
        return(rv)
    })
}

RefSeq_gff.genes <- RefSeq_gff[RefSeq_gff[,3]=="gene",]
RefSeq_gff.genes$Name <- get_Attribute_Field(RefSeq_gff.genes$attributes, "Name", attrsep = ";")
RefSeq_gff.genes$ID <- get_Attribute_Field(RefSeq_gff.genes$attributes, "ID", attrsep = ";")
RefSeq_gff.genes$OldLocusTag <- get_Attribute_Field(RefSeq_gff.genes$attributes, "old_locus_tag", attrsep = ";")
#RefSeq_gff.genes$OldLocusTag = make.unique(RefSeq_gff.genes$OldLocusTag)
colnames(RefSeq_gff.genes)

# Rename Hib genes to old locus tag (available at KEGG)
for (i in 1:nrow(RefSeq_gff.genes)){
  if (is.na(RefSeq_gff.genes$OldLocusTag[i])=="FALSE"){
    rownames(Hib_CountData)[which(rownames(Hib_CountData)==RefSeq_gff.genes$Name[i])] = RefSeq_gff.genes$OldLocusTag[i]
  }
}

Hib_CountData = Hib_CountData[ order(row.names(Hib_CountData)), ]

library(stringr)
# Remove ribosomal RNA
Hib_CountData = Hib_CountData[-which(str_detect(rownames(Hib_CountData), "HIB_r")==TRUE),]
countData = Hib_CountData
```

```{r }
Hib_unaligned_CountData <- as.data.frame(read_delim("~/Dropbox/Hib_Data/unaligned_spades_salmon_counts_matrix.tsv", "\t", escape_double = FALSE, trim_ws = TRUE, skip = 0))

rownames(Hib_unaligned_CountData) = Hib_unaligned_CountData$transcript
Hib_unaligned_CountData = Hib_unaligned_CountData[,-1]

library(seqinr)
Hib_unaligned_fasta = read.fasta("~/Dropbox/Hib_Data/Hib_genes_b2g_HI_Supragenome.fasta", as.string=TRUE)

Hib_unaligned_CountData = Hib_unaligned_CountData[rownames(Hib_unaligned_CountData) %in% names(Hib_unaligned_fasta),]

rownames(Hib_unaligned_CountData)[rownames(Hib_unaligned_CountData)=="NODE_15_length_7629_cov_3174.55_g0_i1"] = "H733_0475"

countData = rbind(Hib_CountData, Hib_unaligned_CountData)
```

# Sets design
```{r }
HibDesign1 = data.frame(row.names = colnames( countData ), condition = rep(c("S01", "S02", "S03", "S04", "S05", "S06"), each = 4))
```

# Least significantly DE genes based on a first-pass DE analysis performed prior to RUVg normalization.
```{r, message=FALSE}
library("RUVSeq")
library("EDASeq")
library("edgeR")
group=as.factor(HibDesign1$condition)
design <- model.matrix(~0 + group)
colnames(design) <- levels(group)
rownames(design) = colnames(countData)

d <- DGEList(counts=countData, lib.size = colSums(countData), group=group)

myCPM = cpm(d)
cpm(1, mean(d$samples$lib.size))
CPMthershold = round(cpm(1, mean(d$samples$lib.size)), digits=2)
keep <- rowSums(cpm(d)> as.vector(CPMthershold)) >= 4
filtered <- d[keep,]
summary(keep)
dim(filtered )

# d <- DGEList(counts=filtered$counts, lib.size = colSums(filtered$counts), group=group)
d <- calcNormFactors(d , method="TMM")
d <- estimateDisp(d , design , tagwise=TRUE, robust=TRUE)
fit <- glmFit(d , design)
lrt <- glmLRT(fit ,coef=2:6)
top <-   topTags(lrt, n=nrow(d))$table
empirical <-   rownames(filtered)[which(!(rownames(filtered) %in% rownames(top)[1:2000]))]
```

# Here, we consider all but the top 500 genes as ranked by edgeR p-values
```{r }
# The RUVg function  returns  two  pieces  of  information:
# the estimated factors of unwanted variation nd the normalized counts obtained by regressing the original counts on the unwanted factors
# The normalized values are stored in the normalizedCounts slot
set2 <- RUVg(as.matrix(d), empirical, k=1)
```

```{r}
col.cell <- c("purple","orange", "blue", "red", "green", "black", "cyan", "darkgoldenrod", "darkorchid", "darkslategray")[as.factor(HibDesign1$condition)]
# plotRLE creates relative log expression (RLE) plot, initially proposed to measure the overall quality of a dataset 
# plotRLE can also be used to visualize the presence of unwanted batch effects in the data
par(mfrow=c(1,2), mar=c(8.1,4.1,4.1,2.1))
plotRLE(as.matrix(filtered$counts), outline=FALSE, ylim=c(-4, 4), col=col.cell, main = "Samples with \n unwanted variation", las=2, cex.axis = 0.8, style = "full",
       outlier.alpha = 0.1, outlier.shape = 3, outlier.size = 0, legend = TRUE)
legend("topright", legend = levels(group), col=unique(col.cell), ncol=2, cex=0.7, border = "black", fill = unique(col.cell))
mtext(side = 1, text = "Samples", line = 7)
mtext(side = 2, text = "Relative Log Expression (RLE)", line = 1.5)

plotRLE(set2$normalizedCounts, outline=FALSE, ylim=c(-4, 4), col=col.cell, main = "Samples without \n unwanted variation", las=2, cex.axis = 0.8, style = "full",
       outlier.alpha = 0.1, outlier.shape = 3, outlier.size = 0)
legend("topright", legend = levels(group), col=unique(col.cell), ncol=2, cex=0.7, border = "black", fill = unique(col.cell))
mtext(side = 1, text = "Samples", line = 7)
mtext(side = 2, text = "Relative Log Expression (RLE)", line = 1.5) 
```

# Set design, removing batch effect
# Create the design matrix
```{r}
# HibDesign2 = data.frame(row.names = colnames( countData ), condition = HibDesign1$condition,  set2$W)
# design = model.matrix(~0+group+set2$W, data = HibDesign2)
# colnames(design) <- c(levels(group), "W")
# group=as.factor(HibDesign1$condition)

HibDesign2 = data.frame(row.names = colnames( set2$normalizedCounts ), condition = HibDesign1$condition)
design = model.matrix(~0+group, data = HibDesign2)
colnames(design) <- c(levels(group))
group=as.factor(HibDesign2$condition)
```

# edgeR
# Coverts the count matrix into an edgeR object
```{r}
d <- DGEList(counts=set2$normalizedCounts, lib.size = colSums(set2$normalizedCounts), group=group)
tpm = cpm(tmm, normalized.lib.sizes=TRUE)
logtmm = cpm(tmm, log=TRUE)
logcounts <- cpm(filtered,log=TRUE)
effectivelib = tmm$samples$lib.size * tmm$samples$norm.factors
disp <- estimateDisp(tmm, design , tagwise=TRUE, robust=TRUE)
fit <- glmFit(disp , design)
nSets = 15
contr.matrix <- makeContrasts(con1 = S02 - S01,
                              con2 = S03 - S01,
                              con3 = S04 - S01,
                              con4 = S05 - S01,
                              con5 = S06 - S01,
                              con6 = S03 - S02,
                              con7 = S04 - S02,
                              con8 = S05 - S02,
                              con9 = S06 - S02,
                              con10 = S04 - S03,
                              con11 = S05 - S03,
                              con12 = S06 - S03,
                              con13 = S05 - S04,
                              con14 = S06 - S04,
                              con15 = S06 - S05,
                              levels = c("S01", "S02", "S03", "S04", "S05", "S06"))

LRT = topDE = topDE1 = topDE2 = nDEGs1 = edgeRDEnames = list()
for (i in 1:nSets){LRT[[i]] = glmLRT(fit, contrast=contr.matrix[,i])}
for (i in 1:nSets){topDE[[i]] = topTags(LRT[[i]] , n=nrow(LRT[[i]]), sort.by = "p.value")}
for (i in 1:nSets){topDE1[[i]] = topDE[[i]]$table[which(abs(topDE[[i]]$table$logFC)>=LFC),]}
for (i in 1:nSets){topDE2[[i]] = topDE1[[i]][which(topDE1[[i]]$FDR<THRESHOLD),]}
for (i in 1:nSets){topDE2[[i]]$edgeR_pi_value = abs(topDE2[[i]]$logFC)*(-1)*log10(topDE2[[i]]$FDR)}
for (i in 1:nSets){topDE2[[i]] = topDE2[[i]][order(topDE2[[i]]$LR, decreasing=TRUE),]}
for (i in 1:nSets){topDE2[[i]]$Gene = rownames(topDE2[[i]])}
for (i in 1:nSets){edgeRDEnames[[i]] = rownames(topDE2[[i]])}
for (i in 1:nSets){nDEGs1[[i]] = length(edgeRDEnames[[i]])}
nDEGs1

is.de = vector(mode = "list", length = nSets)
for (i in 1:nSets){is.de[[i]] = decideTestsDGE(LRT[[i]] , p.value=THRESHOLD)}
DEGsummary = vector(mode = "list", length = nSets)
for (i in 1:nSets){DEGsummary[[i]] = summary(is.de[[i]])}
```


```{r, fig.width=10, fig.height=7}
library("clusterProfiler")
#comparecluster
#data(gcSample)
#xx = compareCluster(gcSample, fun="enrichKEGG", organism="hsa", pvalueCutoff=0.1)

edgeR_DEGs = list(S01xS02 = edgeRDEnames[[1]],
                  S01xS03 = edgeRDEnames[[2]],
                  S01xS04 = edgeRDEnames[[3]],
                  S01xS05 = edgeRDEnames[[4]],
                  S01xS06 = edgeRDEnames[[5]],
                  S02xS03 = edgeRDEnames[[6]],
                  S02xS04 = edgeRDEnames[[7]],
                  S02xS05 = edgeRDEnames[[8]],
                  S02xS06 = edgeRDEnames[[9]], 
                  S03xS04 = edgeRDEnames[[10]], 
                  S03xS05 = edgeRDEnames[[11]], 
                  S03xS06 = edgeRDEnames[[12]], 
                  S04xS05 = edgeRDEnames[[13]], 
                  S04xS06 = edgeRDEnames[[14]], 
                  S05xS06 = edgeRDEnames[[15]])

ee = compareCluster(edgeR_DEGs, fun="enrichKEGG", organism="hiu", pvalueCutoff=0.05)

KEGG_Enrichment_Results_edgeR = data.frame(ee)

mar=c(5.1,4.1,4.1,2.1)
plot(ee, type="dot", font.size = 8, includeAll=TRUE, showCategory=100)

```
